File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_1022_Sum_of_Root_To_Leaf_Binary_Numbers.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_1022_Sum_of_Root_To_Leaf_Binary_Numbers {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private int res;

    public int sumRootToLeaf(TreeNode root) {
        res = 0;
        help(root, 0);
        return res;
    }

    private void help(TreeNode root, int pre) {
        if (root == null) {
            return;
        }
        pre = (pre << 1) | root.val;
        if (root.left != null || root.right != null) {
            help(root.left, pre);
            help(root.right, pre);
        } else {
            res += pre;
        }
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_1022_Sum_of_Root_To_Leaf_Binary_Numbers.cc

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  int sumRootToLeaf(TreeNode* root) {
    int res = 0;
    help(root, 0, res);
    return res;
  }

  void help(TreeNode* root, int pre, int& res) {
    if (!root) {
      return;
    }
    pre = (pre << 1) | root->val;
    if (root->left || root->right) {
      help(root->left, pre, res);
      help(root->right, pre, res);
    } else {
      res += pre;
    }
  }
};
